### ISSUE TITLE
Enabling GPU compromises learning performance

### ISSUE THREAD

**Issue by jimfleming** [google-research/batch-ppo#10]
I've run several trials with TF Agents and found that enabling the GPU through the `use_gpu` configuration flag stalls or inhibits task convergence. Any help troubleshooting this would be appreciated.

The problem seems to exist in all environments but is most prominent in the pendulum task. (Plots below.)

With GPU (4 runs):

<img width="866" alt="screen shot 2017-10-03 at 2 18 04 pm" src="https://user-images.githubusercontent.com/226553/31149716-bd1efaac-a845-11e7-8de5-7bc4da40f1b7.png">

Without GPU (4 runs):

<img width="870" alt="screen shot 2017-10-03 at 2 20 14 pm" src="https://user-images.githubusercontent.com/226553/31149793-ff536b4c-a845-11e7-9f72-174698861e42.png">

These runs were generated with a fresh clone of the TF Agents repo as of this morning but previous versions showed similar results. The only difference between the two graphs is the use of the GPU.

It's also about 3x slower to use the GPU on the pendulum task but I suspect that's due to the relatively small size of the network vs the cost of data transfer to the GPU.

Also tested with:

- CUDNN 6 and 5
- TensorFlow 1.3.0 and 1.2.1
- Both tensorflow and tensorflow-gpu packages (no apparent difference b/w these two for CPU)

(This issue may be related to #8?)

cc @danijar 

EDIT:
GPU run logs: https://gist.github.com/jimfleming/0a163522f02ef9411a5b478099321497
CPU-only run logs: https://gist.github.com/jimfleming/e1eaafb720ee1ee969ea2f4a879ab17b

**Comment by danijar** [comment]
Thanks for the detailed bug report!

Could you print the value of `use_gpu` in algorithm.py (both [line 70][line-70] and [line 547][line-547]), please? That way we can see if all instances of the network indeed get placed on the GPU.

You're right about the data transfer being the bottleneck with small networks, so that GPU only pays off for larger networks.

[line-70]: https://github.com/tensorflow/agents/blob/104a68bf9e61756f173452e1a339b4ddc121e8c5/agents/ppo/algorithm.py#L70
[line-547]: https://github.com/tensorflow/agents/blob/104a68bf9e61756f173452e1a339b4ddc121e8c5/agents/ppo/algorithm.py#L547

**Comment by jimfleming** [comment]
New logs: https://gist.github.com/jimfleming/c347d792818799f998e3219fdfadbcf5

Relevant lines look like:

`PPOAlgorithm __init__ use_gpu [u'/gpu:0']` (x1)

and:

`PPOAlgorithm _network use_gpu [u'/gpu:0']` (x6)

Python's semantics here are weird but I verified that the list is evaluated as `True`. The device is set to `'/gpu:0'`.

**Comment by danijar** [comment]
Thanks for verifying. This seems like it could be a TensorFlow bug since the only thing that changes between the runs is the device assignment.

@ebrevdo Do you have an idea how to debug this further? Do you think it might be related to b/63591557?

**Comment by wookayin** [comment]
Do you guys have any suspect on this? I can reproduce this issue on the latest TensorFlow (1.4.0-rc1) as well. And in terms of wall time gpu versions are always slower.

**Comment by wookayin** [comment]
I have checked into this a bit more, and found that there is some numerical stability issue (like tensorflow/tensorflow#3775).

When we look at the summary, the (scalar) values of

- end_episode/cond/cond/training/adjust_penalty/kl_change
- end_episode/cond/cond/training/adjust_penalty/penalty
- end_episode/cond/cond/training/update_policy/scan_1/while/policy_loss/avg_policy_loss
- ... (and something more)

vanishes to zero on GPU mode, while it has some non-zero values on CPU mode. 

![numerical instability](https://user-images.githubusercontent.com/1009873/32242996-a0c637d8-be4a-11e7-97c7-32912df53f01.png)

Due to this numerical instability, the model seems to be getting no proper loss/gradient signal (the gradient_norm also diverges). I think the computation of KL divergence term is in charge of this. For example, `kl_change` involves `diag_normal_kl()`:

```python
kl_change = tf.reduce_mean(self._mask(utility.diag_normal_kl(
    old_mean, old_logstd, network.mean, network.logstd), length))
```

Note that `diag_normal_kl()` relies on reduction operators, [which is known to be numerically instable on GPUs](https://github.com/tensorflow/tensorflow/issues/3775). So I simply enforced that these operations to run on CPU:

```diff
def diag_normal_kl(mean0, logstd0, mean1, logstd1):
  """Epirical KL divergence of two normals with diagonal covariance."""
+  with tf.device('/cpu:0'):
    logstd0_2, logstd1_2 = 2 * logstd0, 2 * logstd1
    return 0.5 * (
        tf.reduce_sum(tf.exp(logstd0_2 - logstd1_2), -1) +
        tf.reduce_sum((mean1 - mean0) ** 2 / tf.exp(logstd1_2), -1) +
        tf.reduce_sum(logstd1_2, -1) - tf.reduce_sum(logstd0_2, -1) -
        mean0.shape[-1].value)
```

and ~then the training behavior is recovered as in CPU-only training~ (while having `use_gpu=True` and the rest of the graph is placed on GPUs)!

<img src="https://user-images.githubusercontent.com/1009873/32243477-b39c4130-be4b-11e7-8305-54d1a293d528.png" width=300 />


Maybe we'll have to find some alternative formulation of KL divergence that is more stable running on GPUs.

UPDATED: Sorry, I don't think I have fully verified the experiment above. Still it vanishes on GPU perhaps because of the other parts with numerical instability; more careful study is needed.

**Comment by wookayin** [comment]
So, without any cpu placement, the implementation should be obviously in this way:
```python
tf.reduce_sum(
    tf.exp(logstd0_2 - logstd1_2) +
    ((mean1 - mean0) ** 2 / tf.exp(logstd1_2)) +
    (logstd1_2 - logstd0_2), axis=-1)
```
and it again works very well on gpu mode. I'll prepare a PR.

**Comment by danijar** [comment]
Thanks for starting to look into this, that's a good pointer. It trains well on GPU on my machine, so this is quite hard to debug for me. Is the KL computation placed on GPU for you? Only the network computation and RNN states should get placed on GPU ([see constructor](https://github.com/tensorflow/agents/blob/04b32690157e7463d4b92df402ddc7127bd68adf/agents/ppo/algorithm.py#L68:L94)).

I'll update the device placement at some point to optimize for CNN performance, and that might also help with this. If you find time, it would be really helpful to figure out which ops are causing the problem: start by putting everything on CPU and move op by op onto the GPU until it fails.

**Comment by alexpashevich** [comment]
I confirm the same problem using tf-nightly-gpu=1.5.0. It looks like that when allocating the network on GPU, the weights of the network are not updated immediately after applying the gradients.

**Comment by danijar** [comment]
@alexpashevich Thanks for the pull request; I just merged it.

@jimfleming I'm closing this issue now. Please confirm if the changes solve your problem and re-open this issue if not.

**Comment by wookayin** [comment]
The PR: #18
